File: Day-04/typescript/zFlxw/day04.ts
import { readFileLines } from '../utils.ts';

export function day04_01() {
    const lines = readFileLines('/day04/input.txt');
    let totalOverlaps = 0;
    for (const line of lines) {
        if (line === '') continue;
        const [first, second] = line.split(',');
        const [firstLowerBound, firstUpperBound] = first
            .split('-')
            .map((val) => parseInt(val));
        const [secondLowerBound, secondUpperBound] = second
            .split('-')
            .map((val) => parseInt(val));

        if (
            (firstLowerBound <= secondLowerBound &&
                firstUpperBound >= secondUpperBound) ||
            (secondLowerBound <= firstLowerBound &&
                secondUpperBound >= firstUpperBound)
        ) {
            totalOverlaps++;
        }
    }

    console.log('1.', totalOverlaps);
}

export function day04_02() {
    const lines = readFileLines('/day04/input.txt');
    let totalOverlaps = 0;
    for (const line of lines) {
        if (line === '') continue;
        const [first, second] = line.split(',');
        const [firstLowerBound, firstUpperBound] = first
            .split('-')
            .map((val) => parseInt(val));
        const [secondLowerBound, secondUpperBound] = second
            .split('-')
            .map((val) => parseInt(val));

        if (
            Math.max(firstLowerBound, secondLowerBound) <=
            Math.min(firstUpperBound, secondUpperBound)
        ) {
            totalOverlaps++;
        }
    }

    console.log('2.', totalOverlaps);
}

File: Day-04/kotlin/unbekannt/day-04.kt
package unbekannt

import java.io.File

fun main(){
    val input =
        File("input-day04.txt").readLines().map { it.split(",").map { it.split("-").let { IntRange(it[0].toInt(), it[1].toInt()) } }.let { it[0] to it[1] } }

    input.count {
        (it.first.contains(it.second.first) && it.first.contains(it.second.last)) || (it.second.contains(it.first.first) && it.second.contains(it.first.last))
    }.printLn()

    input.count {
        it.first.contains(it.second.first) || it.second.contains(it.first.first)
    }.printLn()

}